File: packages/catalog_core/lib/src/registry.dart
import 'fields.dart';

/// Pet registries (Tasso, FINDEFIX, Petlog, …) differ per country and
/// none of them offers an open interface. cat(a)log therefore knows
/// them only as ID Fields carrying a lookup template: a URL with
/// `{value}` where the number goes. Tapping such a value opens the
/// service in the browser — the app itself never calls them.
const lookupPlaceholder = '{value}';

/// A ready-made registry offered when an ID Field is created, and used
/// to recognize a link found on a flier. Only services with a plain
/// GET URL can be listed; everything else is learned from a scan.
typedef RegistryPreset = ({String name, String template});

/// The one registry with a documented, parameter-addressable report
/// page. FINDEFIX, PETMAXX and Europetnet search by POST only, so they
/// cannot be offered — a poster carrying their link teaches the app the
/// template instead (see [learnLookupTemplate]).
const registryPresets = <RegistryPreset>[
  (
    name: 'Tasso',
    template:
        'https://www.tasso.net/Tierregister/Suchmeldungen?snr=$lookupPlaceholder',
  ),
];

/// The URL for looking [value] up in the service of [def], or null when
/// the Field has no template or the value is empty.
String? lookupUrl(FieldDef def, String value) {
  final template = def.lookupUrl;
  if (template == null || value.trim().isEmpty) return null;
  if (!isWebLookup(template)) return null;
  return buildLookupUrl(template, value);
}

/// [template] with the placeholder replaced by the encoded [value].
String buildLookupUrl(String template, String value) => template
    .replaceAll(lookupPlaceholder, Uri.encodeComponent(value.trim()));

/// The identifier [url] carries for [template], or null when the link
/// belongs to another service. Extra query parameters are ignored on
/// both sides — a scanned Tasso link brings `lang` and `lp` along, and
/// the template of course does not.
String? idFromLookupUrl(String template, String url) {
  final pattern = Uri.tryParse(template);
  final scanned = Uri.tryParse(url);
  if (pattern == null || scanned == null) return null;
  if (pattern.host.toLowerCase() != scanned.host.toLowerCase()) return null;

  // The placeholder sits in a query parameter: read that parameter.
  for (final entry in pattern.queryParameters.entries) {
    if (entry.value != lookupPlaceholder) continue;
    if (_path(pattern) != _path(scanned)) return null;
    final value = scanned.queryParameters[entry.key];
    return (value == null || value.isEmpty) ? null : value;
  }

  // Or in a path segment: every other segment must match.
  final want = pattern.pathSegments;
  final have = scanned.pathSegments;
  if (want.length != have.length) return null;
  String? found;
  for (var i = 0; i < want.length; i++) {
    if (want[i] == lookupPlaceholder) {
      found = have[i];
    } else if (want[i] != have[i]) {
      return null;
    }
  }
  return (found == null || found.isEmpty) ? null : found;
}

/// The preset [url] belongs to, with the identifier it carries — the
/// app can then offer to store that number in a Field of that service.
({RegistryPreset preset, String value})? recognizeLookupUrl(String url) {
  for (final preset in registryPresets) {
    final value = idFromLookupUrl(preset.template, url);
    if (value != null) return (preset: preset, value: value);
  }
  return null;
}

/// Whether [template] (or a scanned link) may be opened from an ID
/// value: web links only. Templates travel as synced entries, so a
/// partner's or a poster's `sms:`/`tel:`/app scheme must not become a
/// tap target on every device.
bool isWebLookup(String template) {
  final uri = Uri.tryParse(template);
  return uri != null && (uri.scheme == 'http' || uri.scheme == 'https');
}

/// Turns a scanned [url] into a template by replacing [value] with the
/// placeholder — this is how an unknown registry becomes usable. Null
/// when the URL does not carry the value at all. Volatile extras
/// (language, paging) are dropped: only the identifying parameter
/// survives.
String? learnLookupTemplate(String url, String value) {
  final scanned = Uri.tryParse(url);
  final wanted = normalizeId(value);
  if (scanned == null || wanted.isEmpty || !isWebLookup(url)) return null;

  for (final entry in scanned.queryParameters.entries) {
    if (normalizeId(entry.value) != wanted) continue;
    return Uri(
      scheme: scanned.scheme,
      host: scanned.host,
      port: scanned.hasPort ? scanned.port : null,
      path: scanned.path,
      queryParameters: {entry.key: lookupPlaceholder},
    ).toString().replaceAll(
        Uri.encodeQueryComponent(lookupPlaceholder), lookupPlaceholder);
  }

  final segments = [...scanned.pathSegments];
  for (var i = 0; i < segments.length; i++) {
    if (normalizeId(segments[i]) != wanted) continue;
    segments[i] = lookupPlaceholder;
    return Uri(
      scheme: scanned.scheme,
      host: scanned.host,
      port: scanned.hasPort ? scanned.port : null,
      pathSegments: segments,
    ).toString().replaceAll(
        Uri.encodeComponent(lookupPlaceholder), lookupPlaceholder);
  }
  return null;
}

/// The parts of [url] that could be the identifier — query parameters
/// and path segments that carry digits. Offered when a link belongs to
/// a registry the app does not know yet, so the user can point at the
/// number instead of typing a template.
List<({String where, String value})> idCandidatesIn(String url) {
  final uri = Uri.tryParse(url);
  if (uri == null) return const [];
  bool looksLikeId(String v) =>
      v.length >= 3 && v.length <= 40 && RegExp(r'\d').hasMatch(v);
  return [
    for (final e in uri.queryParameters.entries)
      if (looksLikeId(e.value)) (where: e.key, value: e.value),
    for (final segment in uri.pathSegments)
      if (looksLikeId(segment)) (where: '/', value: segment),
  ];
}

/// Every http(s) link in a block of text — flier OCR delivers them
/// mid-sentence, often without a scheme.
List<String> urlsIn(String text) {
  final result = <String>[];
  final pattern = RegExp(
      r'(?:https?://|www\.)[^\s<>"' r"'" r']+',
      caseSensitive: false);
  for (final m in pattern.allMatches(text)) {
    var url = m.group(0)!;
    // Trailing sentence punctuation is not part of the link.
    url = url.replaceAll(RegExp(r'[.,;:)\]]+$'), '');
    if (!url.toLowerCase().startsWith('http')) url = 'https://$url';
    if (!result.contains(url)) result.add(url);
  }
  return result;
}

String _path(Uri uri) {
  final path = uri.path;
  if (path.length > 1 && path.endsWith('/')) {
    return path.substring(0, path.length - 1);
  }
  return path;
}

File: desktop/crates/catlog-core/src/registry.rs
//! Pet registries differ per country and none offers an open interface.
//! cat(a)log knows them only as ID Fields carrying a lookup template: a
//! URL with `{value}` where the number goes. The app never calls them.

use crate::fields::{FieldDef, is_web_lookup};

pub const LOOKUP_PLACEHOLDER: &str = "{value}";

/// A ready-made registry offered when an ID Field is created.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct RegistryPreset {
    pub name: &'static str,
    pub template: &'static str,
}

/// The registries with a documented, parameter-addressable report page.
pub const REGISTRY_PRESETS: [RegistryPreset; 1] = [RegistryPreset {
    name: "Tasso",
    template: "https://www.tasso.net/Tierregister/Suchmeldungen?snr={value}",
}];

/// `value` the way a URL component carries it.
pub fn encode_component(value: &str) -> String {
    let mut out = String::new();
    for b in value.trim().bytes() {
        match b {
            b'A'..=b'Z'
            | b'a'..=b'z'
            | b'0'..=b'9'
            | b'-'
            | b'_'
            | b'.'
            | b'!'
            | b'~'
            | b'*'
            | b'\''
            | b'('
            | b')' => out.push(b as char),
            _ => out.push_str(&format!("%{b:02X}")),
        }
    }
    out
}

/// `template` with the placeholder replaced by the encoded `value`.
pub fn build_lookup_url(template: &str, value: &str) -> String {
    template.replace(LOOKUP_PLACEHOLDER, &encode_component(value))
}

/// The URL for looking `value` up in the service of `def`, or none when
/// the Field has no template or the value is empty.
pub fn lookup_url(def: &FieldDef, value: &str) -> Option<String> {
    let template = def.lookup_url.as_deref()?;
    if value.trim().is_empty() || !is_web_lookup(template) {
        return None;
    }
    Some(build_lookup_url(template, value))
}

/// A URL taken apart: host, path segments, query pairs.
type UrlParts = (String, Vec<String>, Vec<(String, String)>);

fn split_url(url: &str) -> Option<UrlParts> {
    let rest = url.split_once("://")?.1;
    let (host_path, query) = match rest.split_once('?') {
        Some((hp, q)) => (hp, Some(q)),
        None => (rest, None),
    };
    let (host, path) = match host_path.split_once('/') {
        Some((h, p)) => (h, p),
        None => (host_path, ""),
    };
    let segments: Vec<String> = path
        .split('/')
        .filter(|s| !s.is_empty())
        .map(String::from)
        .collect();
    let params: Vec<(String, String)> = query
        .unwrap_or("")
        .split('&')
        .filter(|p| !p.is_empty())
        .map(|p| match p.split_once('=') {
            Some((k, v)) => (k.to_string(), v.to_string()),
            None => (p.to_string(), String::new()),
        })
        .collect();
    Some((host.to_lowercase(), segments, params))
}

/// The identifier `url` carries for `template`, or none when the link
/// belongs to another service. Extra query parameters are ignored.
pub fn id_from_lookup_url(template: &str, url: &str) -> Option<String> {
    let (want_host, want_path, want_params) = split_url(template)?;
    let (have_host, have_path, have_params) = split_url(url)?;
    if want_host != have_host {
        return None;
    }
    for (key, value) in &want_params {
        if value != LOOKUP_PLACEHOLDER {
            continue;
        }
        if want_path != have_path {
            return None;
        }
        let found = have_params
            .iter()
            .find(|(k, _)| k == key)
            .map(|(_, v)| v.clone())?;
        return (!found.is_empty()).then_some(found);
    }
    if want_path.len() != have_path.len() {
        return None;
    }
    let mut found: Option<String> = None;
    for (w, h) in want_path.iter().zip(&have_path) {
        if w == LOOKUP_PLACEHOLDER {
            found = Some(h.clone());
        } else if w != h {
            return None;
        }
    }
    found.filter(|f| !f.is_empty())
}

/// The preset `url` belongs to, with the identifier it carries.
pub fn recognize_lookup_url(url: &str) -> Option<(RegistryPreset, String)> {
    REGISTRY_PRESETS
        .iter()
        .find_map(|p| id_from_lookup_url(p.template, url).map(|v| (*p, v)))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::fields::{FieldScope, FieldType, IdDisplay};

    fn def(template: Option<&str>) -> FieldDef {
        FieldDef {
            id: "fielddef:chipid".into(),
            slug: "chipid".into(),
            name: "Chip ID".into(),
            field_type: FieldType::Id,
            scope: FieldScope::Cat,
            options: vec![],
            id_display: IdDisplay::Barcode,
            lookup_url: template.map(String::from),
            dimension: None,
            extra_options: Default::default(),
        }
    }

    #[test]
    fn lookup_links_carry_the_encoded_value() {
        let tasso = def(Some(REGISTRY_PRESETS[0].template));
        assert_eq!(
            lookup_url(&tasso, " 276 098 ").unwrap(),
            "https://www.tasso.net/Tierregister/Suchmeldungen?snr=276%20098"
        );
        assert!(lookup_url(&tasso, "  ").is_none());
        assert!(lookup_url(&def(None), "1").is_none());
        assert!(lookup_url(&def(Some("ftp://x/{value}")), "1").is_none());
        assert_eq!(
            encode_component("a-b_c.d~e!f*g'h(i)j"),
            "a-b_c.d~e!f*g'h(i)j"
        );
        assert_eq!(encode_component("ä/?"), "%C3%A4%2F%3F");
    }

    #[test]
    fn a_scanned_link_gives_its_identifier_back() {
        let t = REGISTRY_PRESETS[0].template;
        assert_eq!(
            id_from_lookup_url(
                t,
                "https://www.tasso.net/Tierregister/Suchmeldungen?lang=de&snr=123&lp=1"
            )
            .as_deref(),
            Some("123")
        );
        assert!(
            id_from_lookup_url(t, "https://www.TASSO.net/Tierregister/Suchmeldungen?snr=")
                .is_none()
        );
        assert!(id_from_lookup_url(t, "https://www.tasso.net/Other?snr=1").is_none());
        assert!(
            id_from_lookup_url(t, "https://other.net/Tierregister/Suchmeldungen?snr=1").is_none()
        );
        assert!(id_from_lookup_url(t, "junk").is_none());
        let path = "https://reg.example/id/{value}/report";
        assert_eq!(
            id_from_lookup_url(path, "https://reg.example/id/42/report").as_deref(),
            Some("42")
        );
        assert!(id_from_lookup_url(path, "https://reg.example/id/42").is_none());
        assert!(id_from_lookup_url(path, "https://reg.example/no/42/report").is_none());
        assert!(id_from_lookup_url("https://reg.example/a/b", "https://reg.example/a/b").is_none());
        let (preset, value) =
            recognize_lookup_url("https://www.tasso.net/Tierregister/Suchmeldungen?snr=9").unwrap();
        assert_eq!((preset.name, value.as_str()), ("Tasso", "9"));
        assert!(recognize_lookup_url("https://nowhere.example/").is_none());
    }
}
